Initial work towards demonstration code and documentation (#1)
CU-DBMI/example-hpc-alpine-python: README.md

Reviewer: This is probably work for a later date, but I could imagine splitting this README into a number of smaller tutorials on:

- submitting jobs to Slurm
- handling environments via Anaconda
- managing data on Alpine
  - e.g., the home, project, and scratch folders and their intended uses/limitations, and how to get data in and out (via Globus)

and retaining this README as an overview of the system and links to those tutorials.

I could also see keeping the value of it all in one file, too, so up to you, really.

Author: Thanks for this - I agree that we should probably have deeper or more specific tutorials on these topics. Candidly I thrashed a bit over whether to include warnings or links to further guidance on those topics to help keep this from getting too long. With this article I sought to create a Pareto principle 20% (of knowledge / training / code) which could offer high value return for the audience's time spent (I don't know if we reached 80%! 🙂 ). At the edges of the content for things which we don't cover my hope is that this at least ignites an eagerness on the audience's part to customize or reach a bit further than we go here.

Other ideas on the content might be that it is used as a short workshop reference to help assist with various issues that could come up along the way; there are a many gaps and questions I can see coming up along the way. Maybe an async way to assist here knowing that schedules could be difficult would be a screenshared recording adjacent to this to help walk through things which is also complemented by the in-depth / focused articles you mentioned.

Author: Created #2 to expand here based on what you suggested, resolving here for now.

            check_1["✅"]
        end
        subgraph time[" &nbsp; 🕑 Long-running Jobs &nbsp;"]
            check_2["✅"]
        end
        subgraph collaborations[" &nbsp; 👥 Collaborations &nbsp;"]
            check_3["✅"]
        end
    end

style alpine fill:#ffffff,stroke:#444444;
style check_1 fill:#D1FAE5,stroke:#D1FAE5;
style check_2 fill:#D1FAE5,stroke:#D1FAE5;
style check_3 fill:#D1FAE5,stroke:#D1FAE5;
style alpine fill:#ffffff,stroke:#444444;
```

_Diagram showing common benefits of Alpine and HPC clusters._

Alpine is a [High Performance Compute (HPC) cluster](https://en.wikipedia.org/wiki/High-performance_computing).
HPC environments provide shared computer hardware resources like [memory](https://en.wikipedia.org/wiki/Computer_memory), [CPU](https://en.wikipedia.org/wiki/Central_processing_unit), [GPU](https://en.wikipedia.org/wiki/Graphics_processing_unit) or others to run performance-intensive work.
Reasons for using Alpine might include:

- __Compute resources:__ Leveraging otherwise cost-prohibitive amounts of memory, CPU, GPU, etc. for processing data.
- __Long-running jobs:__ Completing long-running processes which may take hours or days to complete.

> - __Collaborations:__ Sharing a single implementation environment for reproducibility within a group (avoiding "works on my machine").

Reviewer 2: `"works on my machine" inconsistency issues)`.

Author: Thanks! A lot more descriptive.

    subgraph alpine["🖥️ Alpine"]
        slurm["🗓️ Slurm"]
        subgraph login_nodes["🔑 Login node(s)"]
            acompile["📄 acompile"]
            slurm_cmd["⌨️ Slurm cmd's"]
        end
        subgraph compute_nodes["⚙️ Compute node(s)"]
            hardware["⚙️ Compute\nResources"]
            modules["💿 Software via\nmodules pkg"]
        end

        subgraph storage["📂 Storage"]
            local_storage["📁 Local Storage\n(sometimes temporary)"]
        end
    end
    remote_storage["📁 External Storage\n(user specified / configured)"]

    users --> | preconfigured\nSlurm access| acompile --> slurm
    users --> | direct access | slurm_cmd --> slurm
    slurm --> |"schedules\n(shared) use of"| hardware
    slurm --> | provides\naccess to| modules --> | which may\n leverage| hardware
    hardware --> | may deliver\nresults to| local_storage
    hardware --> | or deliver\nresults to| remote_storage

style alpine fill:#ffffff,stroke:#444444;
style slurm fill:#F0F9FF,stroke:#075985;
```

_Diagram showing high-level user workflow and Alpine components._

Alpine's compute resources are used through compute nodes in a system called [Slurm](https://github.com/SchedMD/slurm).

> Slurm helps coordinate shared and configurable access to the compute resources.

Reviewer: Feel free to take this as-is, change it, or leave it out, but I think diving in a bit deeper into what Slurm is here would make concepts later in the document easier to explain. Here's my attempt; new terms are bolded for emphasis the around the first time they're defined:

```suggestion
Slurm is a system that a large number of users to run jobs on a cluster of computers; the system figures out how to use all the computers in the cluster to execute all the user's jobs fairly (i.e., giving each user approximately equal time and resources on the cluster). A *job* is a request to run something, e.g. a bash script or a program, along with specifications about how much RAM and CPU it needs, how long it can run, and how it should be executed.

Slurm's role in general is to take in a job (submitted via the `sbatch` command) and put it into a *queue* (also called a "partition" in Slurm). For each job in the queue, Slurm constantly tries to find a computer in the cluster with enough resources to run that job, then when an available computer is found runs the program the job specifies on that computer. As the program runs, Slurm records its output to files and finally reports the program's exit status (either completed or failed) back to the job manager.

Importantly, jobs can either be marked as *interactive* or *batch*. When you submit an interactive job, `sbatch` will pause while waiting for the job to start and then connect you to the program, so you can see its output and enter commands in real time. On the other hand, a *batch* job will return immediately; you can see the progress of your job using `squeue`, and you can typically see the output of the job in the folder from which you ran `sbatch` unless you specify otherwise.
```

Author: Cheers, this is fantastic! I'll commit this in.

Data for or from Slurm work may be stored temporarily on local storage or on user-specific external (remote) storage.

> ℹ️ __Wait, what are "nodes"?__
>
> A simplified way to understand the architecture of Slurm on Alpine is through login and compute "nodes" (computers).

> > Login nodes act as a way to prepare and submit processes which will be completed on compute nodes.

Reviewer: You might want to call back to the definition of a job from before, e.g.:

```suggestion
> Login nodes act as a place to prepare and submit jobs which will be completed on compute nodes. Login nodes are never used to execute Slurm jobs, whereas compute nodes are exclusively accessed via a job.
```

Author: Thanks - way clearer! Committing this in.

> Login nodes have limited resource access and are not recommended for running procedures.

One can interact with Slurm on Alpine by use of [Slurm interfaces and directives](https://curc.readthedocs.io/en/latest/clusters/alpine/examples.html).

> A quick way of accessing Alpine resources is through the use of the `acompile` command, which references a script with common Slurm configurations.

Reviewer: Again, a callback to interactive jobs:

```suggestion
A quick way of accessing Alpine resources is through the use of the `acompile` command, which starts an interactive job on a compute node with some typical default parameters for the job. Since `acompile` requests very modest resources (1 hour and 1 CPU core at the time of writing), you'll typically quickly be connected to a compute node. For more intensive or long-lived interactive jobs, consider using `sinteractive`, which allows for more customization [Interactive Jobs](https://curc.readthedocs.io/en/latest/running-jobs/interactive-jobs.html).
```

Author: Thanks, this is a great expansion, committing this in as well.

One can also access Slurm directly through [various commands](https://slurm.schedmd.com/quickstart.html#commands) on Alpine.

Many common software packages are available through the [Modules package](https://github.com/cea-hpc/modules) on Alpine ([UCB RC documentation: The Modules System](https://curc.readthedocs.io/en/latest/compute/modules.html)).

### How does Slurm work?

```mermaid
flowchart LR
    subgraph alpine["🖥️ Alpine"]
        direction LR
        script["📄 Job script"]
        subgraph slurm["🗓️ Slurm"]
            queue["🕑 Queue"]
            processing["⚙️ Processing"]
            completion["✅ Completion\n(or cancellation)"]
        end
    end

    script --> | Submit job\nto Slurm | queue
    queue --> | Scheduled job\nis processed | processing
    processing --> | Processing\ncompletes | completion
    queue --> | Cancellation without\nprocessing| completion

style alpine fill:#ffffff,stroke:#444444;
style slurm fill:#F0F9FF,stroke:#075985;
```

_Diagram showing how Slurm generally works._

Using Alpine effectively involves knowing how to leverage Slurm.
A simplified way to understand how Slurm works is through the following sequence.

> Please note that some steps and additional complexity are obscured for the purposes of providing a basis of understanding.

Reviewer: ```suggestion
Please note that some steps and additional complexity are omitted for the purposes of providing a basis of understanding.
```

These may be distinguished in two ways:

1. __Alpine local storage (sometimes temporary):__ Alpine provides a number of temporary data storage locations for accomplishing your work.
   ⚠️ _Note: some of these locations may be periodically purged and are not a suitable location for long-term data hosting ([see here for more information](https://curc.readthedocs.io/en/latest/additional-resources/policies.html#scratch-file-purge))!_<br>
   Storage locations available ([see this link for full descriptions](https://curc.readthedocs.io/en/latest/compute/filesystems.html)):

   - __Home filesystem:__ 2 GB of backed up space under `/home/$USER` (where `$USER` is your RMACC or Alpine username).
   - __Projects filesystem:__ 250 GB of backed up space under `/projects/$USER` (where `$USER` is your RMACC or Alpine username).
   - __Scratch filesystem:__ 10 TB (10,240 GB) of space __*which is not backed up*__ under `/scratch/alpine/$USER` (where `$USER` is your RMACC or Alpine username).

1. __External / remote storage:__ Users are encouraged to explore external data storage options for long-term hosting.<br>
   Examples may include the following:

>    - __[Petalibrary](https://www.colorado.edu/rc/resources/petalibrary)__: subsidized external storage host from University of Colorado Boulder's Research Computing (requires specific arrangements outside of Alpine).

Reviewer 2: `PetaLibrary`

Reviewer 2: Side note of no consequence, you may want to find+replace `GitHub` to the canon capitalization.

Author: Great points, thank you, will update these in further commits!

### How do I send or receive data on Alpine?

```mermaid
flowchart LR
    external_storage["📁 External Storage\n(user specified / configured)"]
    subgraph alpine["🖥️ Alpine"]
        subgraph storage["📂 Storage"]
            local_storage["📁 Local Storage\n(sometimes temporary)"]
        end
    end

    external_storage --> | send data\nto Alpine | local_storage
    local_storage --> | receive data\nfrom Alpine | external_storage

style alpine fill:#ffffff,stroke:#444444;
```

_Diagram showing external data storage being used to send or receive data on Alpine local storage._

Data may be sent to or gathered from Alpine using a number of different methods.
These may vary contingent on the external data storage being referenced, the code involved, or your group's available resources.
Please reference the following documentation from the University of Colorado Boulder's Research Computing regarding data transfers: [The Compute Environment - Data Transfer](https://curc.readthedocs.io/en/latest/compute/data-transfer.html).
__Please note:__ due to the authentication configuration of Alpine many local or SSH-key based methods are not available for CU Anschutz users.

> As a result, [Globus](https://www.globus.org/) represents one of the best options available (see [3. 📂 Transfer data results](#3-%F0%9F%93%82-transfer-data-results) below).

Reviewer: Awesome; AFAIK Globus is (currently) the recommended mechanism for getting data in and out of Alpine, and it's great to see your detailed tutorial on it later in the text. It may be worth making special note here that Globus can be used both to transfer data into Alpine and out of it, e.g.:

```suggestion
As a result, [Globus](https://www.globus.org/) represents one of the best options available (see [3. 📂 Transfer data results](#3-%F0%9F%93%82-transfer-data-results) below). While the Globus tutorial in this document describes how you can download data from Alpine to your computer, note that you can also use Globus to transfer data to Alpine from your computer.
```

Author: Thanks, great points! Committing this in as well.

### 0. 🔑 Gain Alpine access

First you will need to gain access to Alpine.
This access is provided to members of the University of Colorado Anschutz through [RMACC](https://rmacc.org/) and is separate from other credentials which may be provided by default in your role.
Please see the following guide from the University of Colorado Boulder's Research Computing covering requesting access and generally how this works for members of the University of Colorado Anschutz.

- __RMACC Access to Alpine:__ [https://curc.readthedocs.io/en/latest/access/rmacc.html](https://curc.readthedocs.io/en/latest/access/rmacc.html)

### 1. 🛠️ Prepare code on Alpine

```shell
[username@xsede.org@login-ciX ~]$ cd /projects/$USER
[username@xsede.org@login-ciX username@xsede.org]$ git clone https://github.com/CU-DBMI/example-hpc-alpine-python
Cloning into 'example-hpc-alpine-python'...
... git output ...
[username@xsede.org@login-ciX username@xsede.org]$ ls -l example-hpc-alpine-python
... ls output ...
```

_An example of what this preparation section might look like in your Alpine terminal session._

Next we need to prepare our code within Alpine.

> We do this to balance the fact that we may develop and source control code outside of Alpine and needing to periodically synchronize it with updates.

Reviewer 2: `need`

Author: Thanks! I understood this to mean: "the use of the word 'need' is questionable". I updated this as follows:
```text
Next we will prepare our code within Alpine.
We do this to balance the fact that we may develop and source control code outside of Alpine.
```
